### PR TITLE
Added user restrictions by roles

### LINKED ISSUE #152: Add user restriction

### PR DESCRIPTION
**Pull request by maksym-strus**
Added restriction middleware which helps to protect sensitive data from unauthorized or usual users,  closes #152

### CODE REVIEW

File: __tests__/user/user.mutation.test.js [hunk 1]
```diff
@@ -645,3 +645,96 @@ describe('mutations', () => {
...
+  let lastName;
+  let email;
+
+  beforeAll(() => {
+    firstName = 'Pepo';
+    lastName = 'Markelo';
+    email = 'example@gmail.com';
+    adminId = '9c031d62a3c4909b216e1d86';
+    userId = '5f43af8522155b08109e0304';
+    userToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1ZjQzYWY4NTIyMTU1YjA4MTA5ZTAzMDQiLCJlbWFpbCI6ImV4YW1wbGVAZ21haWwuY29tIiwiaWF0IjoxNTk4Mjk3MjExLCJleHAiOjE1OTkxNjEyMTF9.G0qIuYSb89KNnKy7A2QNTmF6xcsGPhtuodNm8yoUC1s';
```
Comment (Alan18081) [inline __tests__/user/user.mutation.test.js:662]: Please, don't hardcode token, you need to login user in `beforeAll` statement and take the token from the backend

File: __tests__/user/user.query.test.js [hunk 2]
```diff
@@ -297,3 +297,97 @@ describe('queries', () => {
...
+  let userToken;
+  let firstName;
+  let lastName;
+  let adminToken;
+
+  beforeAll(() => {
+    firstName = 'Pepo';
+    lastName = 'Markelo';
+    userId = '5f43af8522155b08109e0304';
+    userToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1ZjQzYWY4NTIyMTU1YjA4MTA5ZTAzMDQiLCJlbWFpbCI6ImV4YW1wbGVAZ21haWwuY29tIiwiaWF0IjoxNTk4Mjk3MjExLCJleHAiOjE1OTkxNjEyMTF9.G0qIuYSb89KNnKy7A2QNTmF6xcsGPhtuodNm8yoUC1s';
```
Comment (Alan18081) [inline __tests__/user/user.query.test.js:311]: Please, don't hardcode token, you need to login user in `beforeAll` statement and take the token from the backend

File: __tests__/user/user.query.test.js [hunk 3]
```diff
@@ -297,3 +297,97 @@ describe('queries', () => {
...
+    adminToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI5YzAzMWQ2MmEzYzQ5MDliMjE2ZTFkODYiLCJlbWFpbCI6ImFkbWluQGdtYWlsLmNvbSIsImlhdCI6MTU5ODI5NzAzOCwiZXhwIjoxNTk5MTYxMDM4fQ.rr9c-ah-X5F0Vy57fuTxMa0gljYt_qc1mIBpI2lis74';
```
Comment (Alan18081) [inline __tests__/user/user.query.test.js:312]: Please, don't hardcode token, you need to login user in `beforeAll` statement and take the token from the backend

File: utils/rules.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  if (context.user) return true;
+  return new UserInputError(USER_NOT_AUTHORIZED, { statusCode: 401 });
+});
+
+const isAdmin = rule()(async (parent, args, context, info) => {
+  if (context.user.role === 'admin') return true;
+  return new UserInputError(INVALID_PERMISSIONS, { statusCode: 403 });
+});
+
+const isTheSameUser = rule()(async (parent, args, context, info) => {
```
Comment (Alan18081) [inline utils/rules.js:19]: Why do we need this kind of check, which operations should have this type of restriction?

File: __tests__/user/user.mutation.test.js (hunk 1 above)
Comment (maksym-strus) [reply __tests__/user/user.mutation.test.js]: fixed, added login user & admin tests for preventing from unexpected bugs and responses

File: __tests__/user/user.query.test.js (hunk 2 above)
Comment (maksym-strus) [reply __tests__/user/user.query.test.js]: fixed, added login user & admin tests for preventing from unexpected bugs and responses

File: __tests__/user/user.query.test.js (hunk 3 above)
Comment (maksym-strus) [reply __tests__/user/user.query.test.js]: fixed, added login user & admin tests for preventing from unexpected bugs and responses

File: utils/rules.js (hunk 4 above)
Comment (maksym-strus) [reply utils/rules.js:19]: We have functions such as getUserById and updateUserById. It would be useful to use them

File: package.json [hunk 5]
```diff
@@ -58,9 +61,9 @@
     "passport-facebook-token": "^3.3.0",
     "passport-google-token": "^0.1.2",
     "rotating-file-stream": "^2.0.2",
+    "start-server-and-test": "^1.11.3",
     "uniqid": "^5.2.0",
-    "util": "^0.12.3",
-    "start-server-and-test": "^1.11.3"
+    "util": "^0.12.3"
```
Comment (Alan18081) [inline package.json:66]: Why do you need this dependency?

Comment (maksym-strus) [reply package.json:66]: It came from the development branch

Comment (Alan18081) [reply package.json:66]: Okay

**Pull request merged**